File: Java/src/main/java/io/github/mihaistreames/afe/algorithms/graphs/DepthFirstSearch.java
package io.github.mihaistreames.afe.algorithms.graphs;

import io.github.mihaistreames.afe.structs.graphs.Edge;
import io.github.mihaistreames.afe.structs.graphs.Graph;
import org.jetbrains.annotations.NotNull;

import java.util.HashMap;
import java.util.Map;
import java.util.Stack;

/**
 * Finds paths in a graph using the Depth-First Search algorithm.
 * <p>
 * Time Complexity: O(V + E) <br>
 * Space Complexity: O(V)
 *
 * @param <T> The type of the vertices.
 */
public class DepthFirstSearch<T> {

    private final Map<T, Boolean> marked;
    private final Map<T, T> edgeTo;
    private final T source;

    /**
     * Performs DFS from a source vertex.
     *
     * @param graph  The graph to search.
     * @param source The source vertex.
     */
    public DepthFirstSearch(@NotNull final Graph<T> graph, final T source) {
        this.marked = new HashMap<>();
        this.edgeTo = new HashMap<>();
        this.source = source;

        for (T vertex : graph.vertices()) {
            marked.put(vertex, false);
        }

        dfs(graph, source);
    }

    /**
     * Checks if there is a path from the source to the given vertex.
     *
     * @param v The vertex.
     * @return true if there is a path, false otherwise.
     */
    public boolean hasPathTo(T v) {
        return marked.getOrDefault(v, false);
    }

    /**
     * Returns the path from the source to the given vertex.
     *
     * @param v The vertex.
     * @return An iterable of the path, or null if no path exists.
     */
    public Iterable<T> pathTo(T v) {
        if (!hasPathTo(v)) return null;

        Stack<T> path = new Stack<>();
        for (T x = v; x != null && !x.equals(source); x = edgeTo.get(x)) {
            path.push(x);
        }
        path.push(source);

        return path;
    }

    private void dfs(@NotNull Graph<T> graph, T v) {
        marked.put(v, true);

        for (Edge<T> edge : graph.adj(v)) {
            T w = edge.other(v); // Get the other vertex from the edge
            if (!marked.get(w)) {
                edgeTo.put(w, v);
                dfs(graph, w);
            }
        }
    }
}

File: Python/afe/algorithms/graphs/depth_first_search.py
from typing import Optional, Iterable, List

from ...structs.graphs import Graph


class DepthFirstSearch:
    """Performs a depth-first search to find paths in a graph.

    Time complexity is O(V + E) where V is vertices and E is edges.
    """

    def __init__(self, graph: Graph, source: int):
        """Initializes and runs the DFS from a source vertex.

        Args:
            graph: The graph to search.
            source: The source vertex (integer).
        """
        self._graph = graph
        self._source = source
        self._marked: List[bool] = [False] * graph.V()
        self._edge_to: List[Optional[int]] = [None] * graph.V()

        graph._validate_vertex(source)
        self._dfs(source)

    def _dfs(self, v: int):
        self._marked[v] = True
        for w in self._graph.adj(v):
            if not self._marked[w]:
                self._edge_to[w] = v
                self._dfs(w)

    def has_path_to(self, v: int) -> bool:
        """Checks if a path exists from the source to vertex v."""
        self._graph._validate_vertex(v)
        return self._marked[v]

    def path_to(self, v: int) -> Optional[Iterable[int]]:
        """Returns a path from the source to vertex v, or None."""
        if not self.has_path_to(v):
            return None

        path: List[int] = []
        x: Optional[int] = v
        while x is not None:
            path.append(x)
            x = self._edge_to[x]

        return reversed(path)
